src/imperative/imperative.cc

```diff
@@ -273,8 +288,126 @@ void Imperative::RecordOp(
+    // a = mx.nd.arange(10)
+    // with dc.context():
+    //     a[:5] = 0
+  }
+  DispatchMode dispatch_mode = DispatchMode::kUndefined;
+  Context default_ctx = Context::CPU();
+  Context ctx = imperative::GetContext(attrs, inputs, outputs, default_ctx);
+  imperative::SetShapeType(ctx, attrs, inputs, outputs, &dispatch_mode);
+
+  nnvm::ObjectPtr node = nnvm::Node::Create();
+  node->inputs.reserve(inputs.size());
+  // Get NodeEntries for inputs
+  for (const NDArray *array : inputs) {
+    // For non-deferred compute arrays, array->deferredcompute_entry_ will be
+    // nullptr. We handle this in in GetDeferredComputeSymbol
+    node->inputs.emplace_back(array->deferredcompute_entry_);
+  }
+  node->attrs = std::move(attrs);
+  // Need to support NameManager in imperative API to better name node->attrs.name
+  node->attrs.name = "node_" + std::to_string(node_count_++);
+  DCInfo::Create(node, inputs, outputs);
+
+  for (uint32_t i = 0; i < outputs.size(); ++i) {
+    outputs[i]->deferredcompute_entry_ = nnvm::NodeEntry{node, i, 0};
+  }
+}
+
+nnvm::Symbol Imperative::GetDeferredComputeSymbol(
+    const std::vector<std::pair<NDArray *, std::string>> &inputs,
+    const std::vector<NDArray *> &outputs
+    ) {
+  Symbol s;
+  s.outputs.reserve(outputs.size());
+  for (NDArray * ndoutput : outputs) {
+    CHECK(!Imperative::DCInfo::IsNone(*ndoutput))
+        << "ValueError: output_arrays for GetDeferredComputeSymbol "
+        << "must have a deferred compute history associated with them.";
+    s.outputs.emplace_back(ndoutput->deferredcompute_entry_);
   }
+  std::unordered_map<NDArray *, nnvm::ObjectPtr> ndinput_to_variable;
+  std::unordered_set<const NDArray *> missing_inputs;
+  auto add_symbol_variables = [&inputs, &ndinput_to_variable,
+                               &missing_inputs](const nnvm::ObjectPtr &node) {
+    if (node == nullptr) {
+      // This (nonexistant) "Node" belongs to an array created outside of deferred compute scope.
+      return;
+    }
+
+    // Check if node has any non-deferred compute inputs
+    for (uint32_t i = 0; i < node->inputs.size(); i++) {
+      nnvm::NodeEntry &node_entry = node->inputs[i];
+      if (node_entry.node == nullptr || node_entry.node->is_variable()) {
+        // Node has non-deferred compute input (nullptr). Find the corresponding
+        // NDArray and create a variable for it. If GetDeferredComputeSymbol has
+        // been called before, a variable already exists and only the name needs
+        // to be updated.
+        Imperative::DCInfo &dcinfo = Imperative::DCInfo::Get(node);
+        const NDArray *array = dcinfo.input_handles_.at(i);
+
+        // Make sure this array is part of GetDeferredComputeSymbol inputs
+        auto is_equal = [array](const std::pair<NDArray *, std::string> &input) {
+          return array == std::get<0>(input);
+        };
+
+        // std::vector<std::pair<NDArray *, std::string>>::iterator input_search =
```

Comment:
nice! thanks for clarifying the type!

Reply:
That one shouldn't have been a comment. I didn't figure out the correct type at the time of writing and used auto as intermediate solution.
Correct type is `std::vector<std::pair<NDArray *, std::string>>::const_iterator` (note the `const`) and is used now.